File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean findTarget(TreeNode root, int k) {
        HashSet<Integer> hs = new HashSet();
        return helper(root, k, hs);
    }
    private boolean helper(TreeNode root, int k, HashSet<Integer> hs) {
        if (root == null) return false;
        int num = root.val;
        int comp = k - num;
        if (hs.contains(num)) {
            return true;
        } else {
            hs.add(comp);
        }
        boolean a = helper(root.left, k, hs);
        boolean b = helper(root.right, k, hs);
        return (a || b);
    }
}

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun findTarget(root: TreeNode?, k: Int): Boolean {
        val compSet = mutableSetOf<Int>()
        return helper(root, k, compSet)
    }
    fun helper(root: TreeNode?, k: Int, compSet: MutableSet<Int>): Boolean {
        if (root == null) return false
        val num = root!!.`val`
        val comp = k-num
        if (num in compSet) {
            return true 
        } else {
            compSet.add(comp)
        }
        val a = (helper(root!!.left, k, compSet))
        val b = (helper(root!!.right, k, compSet))
      return (a || b)
    }
}